File: apps/vertex/src/lib/auth/feide.ts
import { PUBLIC_VERTEX_FEIDE_REDIRECT_URI } from '$env/static/public';
import { OAuth2Client, OAuth2Tokens } from 'arctic';

export type FeideUserInfo = {
	iss: string;
	jti: string;
	aud: string;
	sub: string;
	iat: number;
	exp: number;
	auth_time: number;
	email: string;
	name: string;
	picture: string;
};

const FEIDE_SCOPES = ['openid', 'email', 'profile', 'groups'];

export class Feide {
	#oauthClient: OAuth2Client;
	#authorizationEndpoint = 'https://auth.dataporten.no/oauth/authorization';
	#tokenEndpoint = 'https://auth.dataporten.no/oauth/token';
	#userInfoEndpoint = 'https://auth.dataporten.no/openid/userinfo';

	constructor(clientId: string, clientSecret: string) {
		this.#oauthClient = new OAuth2Client(clientId, clientSecret, PUBLIC_VERTEX_FEIDE_REDIRECT_URI);
	}

	createAuthorizationURL(state: string): URL {
		return this.#oauthClient.createAuthorizationURL(
			this.#authorizationEndpoint,
			state,
			FEIDE_SCOPES
		);
	}

	validateAuthorizationCode(code: string): Promise<OAuth2Tokens> {
		return this.#oauthClient.validateAuthorizationCode(this.#tokenEndpoint, code, null);
	}

	async getUserInfo(accessToken: string): Promise<FeideUserInfo> {
		const response = await fetch(this.#userInfoEndpoint, {
			method: 'GET',
			headers: {
				Authorization: `Bearer ${accessToken}`
			}
		});

		if (!response.ok) {
			throw new Error(`Failed to fetch user info: ${response.statusText}`);
		}

		return await response.json();
	}
}

File: apps/uno/service/auth/feide.go
package auth

import (
	"encoding/json"
	"fmt"
	"io"
	"net/http"
)

const (
	feideAuthEndpoint  = "https://auth.dataporten.no/oauth/authorization"
	feideTokenEndpoint = "https://auth.dataporten.no/oauth/token"
	feideUserEndpoint  = "https://auth.dataporten.no/openid/userinfo"
)

func fetch(url string, headers map[string]string) (res []byte, err error) {
	req, err := http.NewRequest("GET", url, nil)
	if err != nil {
		return nil, fmt.Errorf("failed to create request: %v", err)
	}

	for k, v := range headers {
		req.Header.Set(k, v)
	}

	client := &http.Client{}
	resp, err := client.Do(req)
	if err != nil {
		return nil, fmt.Errorf("request failed: %v", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("failed to fetch user info: %s", resp.Status)
	}

	return io.ReadAll(resp.Body)
}

func getUserInfo(accessToken string) (user FeideUserInfo, err error) {
	res, err := fetch(feideUserEndpoint, map[string]string{
		"Authorization": "Bearer " + accessToken,
	})

	if err != nil {
		return user, err
	}

	if err = json.Unmarshal(res, &user); err != nil {
		return user, err
	}

	return user, err
}
